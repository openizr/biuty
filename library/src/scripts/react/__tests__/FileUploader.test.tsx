/**
 * Copyright (c) Matthieu Jabbour. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import { act, Simulate } from 'react-dom/test-utils';
import UIFileUploader from 'scripts/react/FileUploader';
import { render, unmountComponentAtNode } from 'react-dom';

let container = document.createElement('div');
jest.mock('scripts/helpers/markdown');
jest.mock('scripts/helpers/generateRandomId');

describe('react/UIFileUploader', () => {
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    jest.clearAllMocks();
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    ((container as unknown) as null) = null;
  });

  test('renders correctly - basic', () => {
    act(() => {
      render(<UIFileUploader name="test" modifiers="large" />, container);
    });
    const input = document.querySelector('input') as HTMLInputElement;
    act(() => {
      input.focus();
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      Simulate.change(input, { target: { files: [] } } as any);
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      Simulate.change(input, { target: {} } as any);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with id', () => {
    act(() => {
      render(<UIFileUploader id="test" name="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with label', () => {
    act(() => {
      render(<UIFileUploader name="test" label="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with left icon', () => {
    act(() => {
      render(<UIFileUploader name="test" label="test" icon="star" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with right icon', () => {
    act(() => {
      render(<UIFileUploader name="test" label="test" icon="star" iconPosition="right" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with helper', () => {
    act(() => {
      render(<UIFileUploader name="test" helper="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - multiple', () => {
    act(() => {
      render(<UIFileUploader name="test" multiple />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with placeholder', () => {
    act(() => {
      render(<UIFileUploader name="test" placeholder="test" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with accept', () => {
    act(() => {
      render(<UIFileUploader name="test" accept="image/*" />, container);
    });
    expect(container).toMatchSnapshot();
  });

  test('renders correctly - with listeners', () => {
    const onChange = jest.fn();
    const onFocus = jest.fn();
    act(() => {
      render(<UIFileUploader name="test" onChange={onChange} onFocus={onFocus} />, container);
    });
    const input = document.querySelector('input') as HTMLInputElement;
    act(() => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      Simulate.change(input, { target: { files: ['/path/to/file1.png'] } } as any);
      input.focus();
    });
    expect(container).toMatchSnapshot();
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith(['/path/to/file1.png']);
    expect(onFocus).toHaveBeenCalledTimes(1);
  });
});
