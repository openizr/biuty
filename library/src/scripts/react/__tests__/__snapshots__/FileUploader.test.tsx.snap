// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react/UIFileUploader renders correctly - basic 1`] = `
<div>
  <div
    class="ui-file-uploader ui-file-uploader--large"
  >
    <input
      class="ui-file-uploader__field"
      id="z8a8d7a5zad4"
      name="test"
      type="file"
    />
    <span
      class="ui-file-uploader__files-list"
    />
  </div>
</div>
`;

exports[`react/UIFileUploader renders correctly - multiple 1`] = `
<div>
  <div
    class="ui-file-uploader"
  >
    <input
      class="ui-file-uploader__field"
      id="z8a8d7a5zad4"
      multiple=""
      name="test"
      type="file"
    />
    <span
      class="ui-file-uploader__files-list"
    />
  </div>
</div>
`;

exports[`react/UIFileUploader renders correctly - with helper 1`] = `
<div>
  <div
    class="ui-file-uploader"
  >
    <input
      class="ui-file-uploader__field"
      id="z8a8d7a5zad4"
      name="test"
      type="file"
    />
    <span
      class="ui-file-uploader__files-list"
    />
    <p
      class="ui-file-uploader__helper"
    >
      test
    </p>
  </div>
</div>
`;

exports[`react/UIFileUploader renders correctly - with id 1`] = `
<div>
  <div
    class="ui-file-uploader"
    id="test"
  >
    <input
      class="ui-file-uploader__field"
      id="z8a8d7a5zad4"
      name="test"
      type="file"
    />
    <span
      class="ui-file-uploader__files-list"
    />
  </div>
</div>
`;

exports[`react/UIFileUploader renders correctly - with label 1`] = `
<div>
  <div
    class="ui-file-uploader"
  >
    <label
      class="ui-file-uploader__label"
      for="z8a8d7a5zad4"
    >
      test
    </label>
    <input
      class="ui-file-uploader__field"
      id="z8a8d7a5zad4"
      name="test"
      type="file"
    />
    <span
      class="ui-file-uploader__files-list"
    />
  </div>
</div>
`;

exports[`react/UIFileUploader renders correctly - with left icon 1`] = `
<div>
  <div
    class="ui-file-uploader"
  >
    <label
      class="ui-file-uploader__label"
      for="z8a8d7a5zad4"
    >
      test
    </label>
    <i
      class="ui-button__icon"
    >
      star
    </i>
    <input
      class="ui-file-uploader__field"
      id="z8a8d7a5zad4"
      name="test"
      type="file"
    />
    <span
      class="ui-file-uploader__files-list"
    />
  </div>
</div>
`;

exports[`react/UIFileUploader renders correctly - with listeners 1`] = `
<div>
  <div
    class="ui-file-uploader"
  >
    <input
      class="ui-file-uploader__field"
      id="z8a8d7a5zad4"
      name="test"
      type="file"
    />
    <span
      class="ui-file-uploader__files-list"
    />
  </div>
</div>
`;

exports[`react/UIFileUploader renders correctly - with placeholder 1`] = `
<div>
  <div
    class="ui-file-uploader"
  >
    <input
      class="ui-file-uploader__field"
      id="z8a8d7a5zad4"
      name="test"
      type="file"
    />
    <span
      class="ui-file-uploader__files-list"
    >
      test
    </span>
  </div>
</div>
`;

exports[`react/UIFileUploader renders correctly - with right icon 1`] = `
<div>
  <div
    class="ui-file-uploader"
  >
    <label
      class="ui-file-uploader__label"
      for="z8a8d7a5zad4"
    >
      test
    </label>
    <input
      class="ui-file-uploader__field"
      id="z8a8d7a5zad4"
      name="test"
      type="file"
    />
    <i
      class="ui-button__icon"
    >
      star
    </i>
    <span
      class="ui-file-uploader__files-list"
    />
  </div>
</div>
`;
